x/ibc/light-clients/07-tendermint/types/upgrade.go

```diff
@@ -26,102 +25,129 @@ import (
+	consState, err := GetConsensusState(clientStore, cdc, lastHeight)
 	if err != nil {
-		return nil, nil, sdkerrors.Wrap(err, "could not retrieve consensus state for upgradeHeight")
+		return nil, nil, sdkerrors.Wrap(err, "could not retrieve consensus state for lastHeight")
 	}
 
 	if cs.IsExpired(consState.Timestamp, ctx.BlockTime()) {
 		return nil, nil, sdkerrors.Wrap(clienttypes.ErrInvalidClient, "cannot upgrade an expired client")
 	}
 
-	tmCommittedClient, ok := committedClient.(*ClientState)
-	if !ok {
-		return nil, nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClientType, "upgraded client must be Tendermint client. expected: %T got: %T",
-			&ClientState{}, upgradedClient)
+	// Verify client proof
+	bz, err := codec.MarshalAny(cdc, committedClient)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(clienttypes.ErrInvalidClient, "could not marshal client state: %v", err)
+	}
+	// construct clientState Merkle path
+	upgradeClientPath := constructUpgradeClientMerklePath(cs.UpgradePath[:], lastHeight)
+	if err := merkleProofClient.VerifyMembership(cs.ProofSpecs, consState.GetRoot(), upgradeClientPath, bz); err != nil {
+		return nil, nil, err
 	}
 
+	// Verify consensus state proof
+	bz, err = codec.MarshalAny(cdc, upgradedConsState)
+	if err != nil {
+		return nil, nil, sdkerrors.Wrapf(clienttypes.ErrInvalidConsensus, "could not marshal consensus state: %v", err)
+	}
+	// construct consensus state Merkle path
+	upgradeConsStatePath := constructUpgradeConsStateMerklePath(cs.UpgradePath[:], lastHeight)
+	if err := merkleProofConsState.VerifyMembership(cs.ProofSpecs, consState.GetRoot(), upgradeConsStatePath, bz); err != nil {
+		return nil, nil, err
+	}
+
+	// Construct new client state and consensus state
 	// Relayer chosen client parameters are ignored.
 	// All chain-chosen parameters come from committed client, all client-chosen parameters
 	// come from current client.
-	updatedClientState := NewClientState(
+	newClientState := NewClientState(
 		tmCommittedClient.ChainId, cs.TrustLevel, cs.TrustingPeriod, tmCommittedClient.UnbondingPeriod,
 		cs.MaxClockDrift, tmCommittedClient.LatestHeight, tmCommittedClient.ProofSpecs, tmCommittedClient.UpgradePath,
 		cs.AllowUpdateAfterExpiry, cs.AllowUpdateAfterMisbehaviour,
 	)
 
-	if err := updatedClientState.Validate(); err != nil {
+	if err := newClientState.Validate(); err != nil {
 		return nil, nil, sdkerrors.Wrap(err, "updated client state failed basic validation")
 	}
 
-	if err := merkleProof.VerifyMembership(cs.ProofSpecs, consState.GetRoot(), upgradePath, bz); err != nil {
-		return nil, nil, err
-	}
+	// The new consensus state is merely used as a trusted kernel against which headers on the new
+	// chain can be verified. The root is empty as it cannot be known in advance, thus no proof verification will pass.
+	// The timestamp of the consensus state is also this chain's blocktime. This is because starting up a new chain
```

Comment:
is this safe?

If a chain upgrades at time t, and starts producing blocks at t +1 and this client upgrades at time t + 5, wouldn't updates fail to be processed for t + 1 through t + 4?

Reply:
Yes, my thinking is the gap between when a chain starts producing blocks and when a relayer sends the upgrade tx is bounded by the relayer liveness assumption. So yes, it's possible that some of the early blocks may be in the past relative to our light client. But we may simply wait until the blockchain goes past our light client (time-wise) and then submit blocks from that point on. There's no safety issue here, since we can verify proofs fine against all these future blocks, and if the light client expires this implies the light client would have expired anyway.

The bigger safety issue in my mind is that we use the time committed to in the consensus state on the previous chain and some unforeseen issue happens either during the upgrade or before the upgrade but after the plan has been committed that will either stall the old chain or prevent the new chain from being restarted quickly. This could cause all light clients on the old chain to expire.

Reply:
cc: @cwgoes

Reply:
wouldn't this tendermint [check](https://github.com/tendermint/tendermint/blob/a736530e012c168b60cf4a0f6c6a0bc7036a00f0/light/verifier.go#L170) fail

Upgrade at: t
Untrusted header (first header after update): t + 1
Trusted header (stored consensus state at time of msgupgrade): t + 5

In this case and every case untrusted header is not after trusted header, right?

Reply:
Correct, my point is we can just wait for the first untrusted header that is past the trusted header, and submit that without an issue

Reply:
Sure, this also requires the validator set from the time of the upgrade to the time of the msg upgrade + 1 to still fall within the trust level. This should be well documented, clients with strong trust levels could easily become stuck

would like @cwgoes to ack this approach as well

Reply:
Hmm, shouldn't we allow the upgrading chain to (somehow) specify the timestamp of the new root-of-trust? It would be equivalent to start_time in genesis, for example, which might be a few hours after the old chain stops or otherwise pre-programmed in advance. I don't like using a local time because it means different chains with different local times may behave differently (e.g. one may succeed in following the upgraded client and one may not), which is a violation of the usual light client guarantees. I'd rather we use a timestamp specified (and proved) by the upgrading chain and handle it the same way in all chains following the upgrade.

Does that seem possible? Would we require an additional proof of this timestamp?

Reply:
Hmm how does genesis start time get chosen? Does it get exported in genesis or is it just chosen outside of consensus?

Reply:
Presumably when creating an upgrade a new genesis time, at least, is specified? Is this not already in the state machine?

Reply:
is there no way to set the consensus state during the execution at `plan.Height`? Then we could just use the block time right?

Reply:
Yes, that's possible. I guess we can just enforce that the upgrade occurs successfully within the trusting period of the client. I suppose that's reasonable 🤷‍♂️

Reply:
actually wait nvm, then we cannot prove at `plan.Height` because the light client isn't updated at that height yet

I assume you are going to use the block time when setting the consensus state at `plan.Height -1` and then the upgrade msg just needs to be submitted within the trusting period (the `plan.Height - 1` should be committed even if the upgrade hasn't happened right?)

Reply:
I set the consenus state time on the last height of the old chain: `plan.Height-1`. This is safe and will work so long as the first block of the new chain is submitted to the upgraded client before `lastBlockTime + trustingPeriod`. Otherwise, the upgraded client will have expired.

Thus success of upgrade depends on our local choice of trusting period, but if the upgrade happens reasonably soon we can be confident most clients will successfully upgrade.